runtime/commonMain/src/kotlinx/serialization/encoding/Encoding.kt

```diff
@@ -419,8 +434,11 @@ public interface CompositeEncoder {
 /**
  * Begins a structure, encodes it using the given [block] and ends it.
  */
-public inline fun Encoder.encodeStructure(descriptor: SerialDescriptor, crossinline block: CompositeEncoder.() -> Unit) {
+public inline fun Encoder.encodeStructure(descriptor: SerialDescriptor, block: CompositeEncoder.() -> Unit) {
     val composite = beginStructure(descriptor)
-    composite.block()
-    composite.endStructure(descriptor)
+    try {
+        composite.block()
+    } finally {
```

Comment:
Not sure how much benefit it will bring since we do not give any exception-safety guarantees on encoder/decoder internal state anyway

Reply:
This is done mostly for convenience, e.g. take a look at `AbstractPolymorphicSerializer#deserialize`